odl/deform/linearized.py

```diff
@@ -0,0 +1,325 @@
+    >>> _linear_deform(template, displacement_field)
+    array([ 0.,  0.,  1.,  1.,  0.])
+    """
+    image_pts = template.space.points()
+    for i, vi in enumerate(displacement):
+        image_pts[:, i] += vi.asarray().ravel()
+    return template.interpolation(image_pts.T, out=out, bounds_check=False)
+
+
+class LinDeformFixedTempl(Operator):
+
+    """Operator mapping a displacement field to corresponding deformed template.
+
+    The operator has a fixed template ``I`` and maps a displacement
+    field ``v`` to the new function ``x --> I(x + v(x))``.
+    """
+
+    def __init__(self, template, domain=None):
+        """Initialize a new instance.
+
+        Parameters
+        ----------
+        template : `DiscreteLpVector` or element-like
+            Fixed template that is to be deformed.
+            If ``domain`` is not given, ``template`` must
+            be a `DiscreteLpVector`, and the domain of this operator
+            is inferred from ``template.space``. If ``domain`` is
+            given, ``template`` can be anything that is understood
+            by the ``domain[0].element()`` method.
+        domain : product space of `DiscreteLp`, optional
+            Space of displacement fields on which this operator
+            acts, i.e. the operator domain. If not given, it is
+            inferred from ``template.space``.
+
+        Examples
+        --------
+        Create a template and deform it with a given deformation field.
+
+        Where the deformation field is zero we expect to get the same output
+        as the input. In the 4:th point, the deformation is non-zero and hence
+        we expect to get the value of the point 0.2 to the left, that is 1.0.
+
+        >>> import odl
+        >>> space = odl.uniform_discr(0, 1, 5, interp='nearest')
+        >>> template = space.element([0, 0, 1, 0, 0])
+        >>> op = LinDeformFixedTempl(template)
+        >>> disp_field = [[0, 0, 0, -0.2, 0]]
+        >>> print(op(disp_field))
+        [0.0, 0.0, 1.0, 1.0, 0.0]
+
+        The result depends on the chosen interpolation. If we chose 'linear'
+        interpolation and offset the point half the distance between two
+        points, 0.1, we expect to get the mean of the values.
+
+        >>> space = odl.uniform_discr(0, 1, 5, interp='linear')
+        >>> template = space.element([0, 0, 1, 0, 0])
+        >>> op = LinDeformFixedTempl(template)
+        >>> disp_field = [[0, 0, 0, -0.1, 0]]
+        >>> print(op(disp_field))
+        [0.0, 0.0, 1.0, 0.5, 0.0]
+
+        See Also
+        --------
+        LinDeformFixedDisp : Deformation with a fixed displacement.
+        """
+        if domain is None:
+            if not isinstance(template, DiscreteLpVector):
+                raise TypeError('`template` must be a `DiscreteLpVector`'
+                                'instance if `domain` is None, got {!r}'
+                                ''.format(template))
+
+            domain = template.space.tangent_space
```

Comment:
The source for the discussion regarding the notion of tangent space is that ODL lacks a notion of manifold (curved space). In ODL one has either a set or a vector space, the notion of a. manifold does not exist. We have already encountered the need to handle manifolds since data space in tomography is a non-Euclidian manifold. The operations required for points (=tomographic data) on such manifolds could however be rephrased as operations on the underlying coordinate space (products of intervals), ignoring the fact that a single coordinate space is insufficient as an atlas. Now, dealing with deformations requires us yet again to handle points (=vector fields) in curved spaces, i.e. manifolds. One option is to try to recast the operations to the underlying coordinate spaces, e.g. assuming a RKHS. Another is to introduce notions manifold in ODL with a tangent bundle. The latter seems to be quite time consuming though.